wazo_confd/plugins/line/resource.py

```diff
@@ -25,16 +86,64 @@ def get(self):
 
     @required_acl('confd.lines.create')
     def post(self):
-        form = self.schema().load(request.get_json())
+        return super().post()
+
+    def _post(self, item):
+        form = self.schema().load(item)
+
+        endpoint_sip_body = form.pop('endpoint_sip', None)
+        endpoint_sccp_body = form.pop('endpoint_sccp', None)
+        endpoint_custom_body = form.pop('endpoint_custom', None)
+        extensions = form.pop('extensions', None) or []
+
         model = self.model(**form)
         tenant_uuids = self._build_tenant_list({'recurse': True})
         model = self.service.create(model, tenant_uuids)
-        return self.schema().dump(model), 201, self.build_headers(model)
+
+        payload = self.schema().dump(model)
+        payload['extensions'] = []
+
+        if endpoint_sip_body:
+            endpoint_sip, _, _ = self._endpoint_sip_list_resource._post(
+                endpoint_sip_body
+            )
+            self._line_endpoint_sip_association_resource.put(
+                model.id,
+                endpoint_sip['uuid'],
+            )
+            payload['endpoint_sip'] = endpoint_sip
+        elif endpoint_sccp_body:
+            endpoint_sccp, _, _ = self._endpoint_sccp_list_resource._post(
+                endpoint_sccp_body
+            )
+            self._line_endpoint_sccp_association_resource.put(
+                model.id,
+                endpoint_sccp['id'],
+            )
+            payload['endpoint_sccp'] = endpoint_sccp
+        elif endpoint_custom_body:
+            endpoint_custom, _, _ = self._endpoint_custom_list_resource._post(
+                endpoint_custom_body
+            )
+            self._line_endpoint_custom_association_resource.put(
+                model.id,
+                endpoint_custom['id'],
+            )
+            payload['endpoint_custom'] = endpoint_custom
+
+        for extension_body in extensions:
+            extension, _, _ = self._extension_line_list_resource._post(
+                model.id, extension_body
+            )
+            payload['extensions'].append(extension)
```

Comment:
I agree that `service.py` should be the place to do most of this logic.

The current implementation in restful does many things
1. Extracts the payload from the request
2. Deserializes the payload
3. Add the `tenant_uuid` to the payload when appropriate
4. Instantiate the sqlalchemy model from the payload
5. Calls `service.create`
6. Serializes the response
7. Return the response, status code and headers

I could extract steps 4 and move it into the service. That would allow the input to be inspected or modified before instantiating the sqlalchemy model.

I "think" (I would have to try with a resource first) that that would allow me to move all logic from the resources to the services. Also it would avoid the necessity of splitting the `post` in 2 functions in `restful.py`

I would also like to introduce a new class `ServiceHandle` that would reference all services that could be added to the `dependencies` and allow multiple plugins to use a service without having to instantiate the service and import all of its dependencies.

That would be used like that
```diff
diff --git a/wazo_confd/plugins/extension/plugin.py b/wazo_confd/plugins/extension/plugin.py
index 77d5125a..91fce513 100644
--- a/wazo_confd/plugins/extension/plugin.py
+++ b/wazo_confd/plugins/extension/plugin.py
@@ -1,4 +1,4 @@
-# Copyright 2016-2019 The Wazo Authors  (see the AUTHORS file)
+# Copyright 2016-2022 The Wazo Authors  (see the AUTHORS file)
 # SPDX-License-Identifier: GPL-3.0-or-later
 
 from wazo_provd_client import Client as ProvdClient
@@ -12,11 +12,13 @@ class Plugin:
         api = dependencies['api']
         config = dependencies['config']
         token_changed_subscribe = dependencies['token_changed_subscribe']
+        service_handle = dependencies['service_handle']
 
         provd_client = ProvdClient(**config['provd'])
         token_changed_subscribe(provd_client.set_token)
 
         service = build_service(provd_client)
+        service_handle.register('extension', service)
 
         api.add_resource(
             ExtensionItem,
```

```diff
diff --git a/wazo_confd/plugins/line/resource.py b/wazo_confd/plugins/line/resource.py
index 5230bf39..190b7a75 100644
--- a/wazo_confd/plugins/line/resource.py
+++ b/wazo_confd/plugins/line/resource.py
@@ -19,6 +19,10 @@ class LineList(ListResource):
     def build_headers(self, line):
         return {'Location': url_for('lines', id=line.id, _external=True)}
 
+    def __init__(self, line_service, service_handle):
+        super().__init__(line_service)
+        self._extension_service = service_handle.get('extension')
+
     @required_acl('confd.lines.read')
     def get(self):
         return super().get()
```
```diff
diff --git a/wazo_confd/controller.py b/wazo_confd/controller.py
index 23ff58df..de65893e 100644
--- a/wazo_confd/controller.py
+++ b/wazo_confd/controller.py
@@ -24,6 +24,18 @@ from .service_discovery import self_check
 logger = logging.getLogger(__name__)
 
 
+class ServiceHandle:
+
+    def __init__(self):
+        self._services = {}
+
+    def register(self, name, service):
+        self._services[name] = service
+
+    def service(self, name):
+        return self._services[name]
+
+
 class Controller:
     def __init__(self, config):
         self.config = config
@@ -51,6 +63,7 @@ class Controller:
         self.token_renewer.subscribe_to_token_change(
             self.token_status.token_change_callback
         )
+        service_handle = ServiceHandle()
         self.status_aggregator.add_provider(auth.provide_status)
         self.status_aggregator.add_provider(self.token_status.provide_status)
         self.status_aggregator.add_provider(self._bus_consumer.provide_status)
@@ -66,6 +79,7 @@ class Controller:
                 'bus_publisher': self._bus_publisher,
                 'auth_client': auth_client,
                 'pjsip_doc': pjsip_doc,
+                'service_handle': service_handle,
                 'status_aggregator': self.status_aggregator,
             },
         )
```